File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    vis = [0] * 20001
    line = [[] for _ in range(20001)]
    max_len = 0
    
    for value in edge:
        line[value[0]].append(value[1])
        line[value[1]].append(value[0])
    
    Q = deque()
    Q.append(1)
    vis[1] = 1
    
    while Q:
        cur = Q.popleft()
        
        for value in line[cur]:
            np = value
            
            if (vis[np] > 0):
                continue
            
            Q.append(np)
            vis[np] = vis[cur] + 1
            if (vis[np] > max_len):
                max_len = vis[np]
                
    for value in vis:
        if (value == max_len):
            answer += 1
            
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    int last_cnt = 0;
    vector<int> V[20005];
    int vis[20005] = { 0 };
    queue<int> Q;
    
    for (int i = 0; i < edge.size(); i++) {
        V[edge[i][0]].push_back(edge[i][1]);
        V[edge[i][1]].push_back(edge[i][0]);
    }
     
    Q.push(1);
    vis[1] = 1;
        
    while (!Q.empty()) {
        int cur = Q.front(); Q.pop();
        
        for (int i = 0; i < V[cur].size(); i++) {
            if (vis[V[cur][i]])
                continue;
            
            vis[V[cur][i]] = vis[cur] + 1;
            Q.push(V[cur][i]);
            last_cnt = vis[V[cur][i]];
        }
    }
    
    for (int i = 1; i <= n; i++) {
        if (vis[i] == last_cnt)
            answer++;
    }
    
    return answer;
}